File: aoc2023/ocaml_solutions/lib/day01.ml
open Core
open Util

let replacements1 =
  List.(map (range 1 9 ~stop:`inclusive)
          ~f:(fun n -> (string_of_int n, n)))

let replacements2 =
  replacements1 @ [ ("one",   1)
                  ; ("two",   2)
                  ; ("three", 3)
                  ; ("four",  4)
                  ; ("five",  5)
                  ; ("six",   6)
                  ; ("seven", 7)
                  ; ("eight", 8)
                  ; ("nine",  9)
                  ]

let solve (repls : (string * int) list) inp =
  let find_number isIn xs =
    let rec go = function
      | []        -> failwith "impossible"
      | (s :: ss) ->
         match List.find repls ~f:(fun (a, _) -> isIn a s) with
         | None        -> go ss
         | Some (_, b) -> b
    in go xs in
  let first = find_number
                (fun prefix b -> String.is_prefix ~prefix b)
                (String.tails inp)
  and second = find_number
                 (fun suffix b -> String.is_suffix ~suffix b)
                 (List.rev (String.inits inp))
  in first * 10 + second

let day01 =
  let inp = In_channel.read_lines "../../inputs/day01.txt" in
  let find_nums repls = string_of_int @@ sum @@
                          List.map inp ~f:(fun s -> solve repls s)
  in find_nums replacements1
   , find_nums replacements2

File: aoc2023/haskell-solutions/src/Day01.hs
{-# OPTIONS_GHC -Wno-name-shadowing #-}
module Day01 (day01) where

import Util

day01 :: IO (Int, Int)
day01 = do
  input <- lines <$> readFile' "../inputs/day01.txt"
  let findNums repls = sum $ map (solve repls) input
  pure (findNums replacements1, findNums replacements2) -- 54331, 54518

solve :: [(String, Int)] -> String -> Int
solve repls inp =
  let first = findWork isPrefixOf (tails inp)
      last  = findWork isSuffixOf (reverse $ inits inp)
   in 10 * first + last
 where
  findWork :: (String -> String -> Bool) -> [String] -> Int
  findWork isIn = go
   where
    go []       = error "impossible"
    go (s : ss) = case find (\(a, _) -> a `isIn` s) repls of
      Nothing     -> go ss
      Just (_, b) -> b

replacements1 :: [(String, Int)]
replacements1 = [(show n, n) | n <- [1..9 :: Int]]

replacements2 :: [(String, Int)]
replacements2
   = replacements1
  <> [ ("one",   1)
     , ("two",   2)
     , ("three", 3)
     , ("four",  4)
     , ("five",  5)
     , ("six",   6)
     , ("seven", 7)
     , ("eight", 8)
     , ("nine",  9)
     ]
